waku/protocol/v2/waku_filter.nim

```diff
@@ -0,0 +1,96 @@
+import chronos, chronicles
+import ./filter
+import tables
+import libp2p/protocols/pubsub/pubsub,
+       libp2p/protocols/pubsub/pubsubpeer,
+       libp2p/protocols/pubsub/floodsub,
+       libp2p/protocols/pubsub/gossipsub,
+       libp2p/protocols/pubsub/rpc/[messages, protobuf],
+       libp2p/protocols/protocol,
+       libp2p/protobuf/minprotobuf,
+       libp2p/stream/connection
+
+import metrics
+
+import stew/results
+
+const
+  WakuFilterCodec* = "/vac/waku/filter/2.0.0-alpha2"
+
+type
+  ContentFilter* = object
+    topics*: seq[string]
+
+  FilterRPC* = object
```

Comment:
It is not clear to me what are the actual RPC calls, and how the "filtered" messages are flowing from e.g. a full client to a light client. But, I'm probably missing quite some background on this.

Reply:
There's only RPC call to "install" a filter: https://specs.vac.dev/specs/waku/waku-v2.html#content-filtering

For filtered messages flowing to a light client there is only this:

> A node that only sets this field but doesn’t subscribe to any topic SHOULD only get notified when the content subtopic matches. A content subtopic matches when a Wakumessage contentTopic field is the same.

So it should impact logic for how messages are relayed to other nodes. I don't think all of that has been implemented yet, but @decanus can probably elaborate.

Do you think it should be structured differently, or specified in more detail?

Reply:
Well, I had in mind that this would be another protocol with a contentFilter call to pass the contentFilter (say topic interest) to other peers, and then also a call that actually pushes these messages (or a request response if you want to poll).  So the latter would be missing right now.

But from the sentence `A node that receives this RPC SHOULD apply this content filter before relaying.` in the spec, I understand that we actually want to still use PubSub (Gossip or Flood) here but alter the behaviour towards a selective set of peers (the lightnodes).
I guess this would mean to have a new `rpcHandler` that does the same as for flood/gossipsub, except for these peers. Not sure if that is so convenient to do and if current `rpcHandlers` can be still reused in the overriding calls, I'd have to go through to code better. I guess @dryajov could probably answer this much faster.

Reply:
So, I'll do my best to answer given my current understanding on what the requirement here is, but keep in mind that I might be missing pieces of the model.

There are a few things to keep in mind there:

- Pubsub has the ability of filtering messages already, we call them validators and they control weather a message gets propagates or not.
- Pubsub might not be the right protocol for any sort of "light" message exchange.

> Well, I had in mind that this would be another protocol with a contentFilter call to pass the contentFilter (say topic interest) to other peers, and then also a call that actually pushes these messages (or a request response if you want to poll). So the latter would be missing right now.

This makes sense to me too. If what we want is to reduce the load on "light" nodes, then we should exclude them from gossip altogether, and in that case a stand alone protocol such as what @kdeme is describing would make the most sense. Potentially request/response, where the client sends the topics it's interested in and gets back a set of messages for those topics.

> But from the sentence A node that receives this RPC SHOULD apply this content filter before relaying. in the spec, I understand that we actually want to still use PubSub (Gossip or Flood) here but alter the behaviour towards a selective set of peers (the lightnodes).

`A node that receives this RPC SHOULD apply this content filter before relaying` is something that pubsub already can do with the validators, furthermore, validators receive the entire message object, not just the data field, so it's possible to use additional fields such as from and sequence, to filter messages (might not be needed for waku).

> I guess this would mean to have a new rpcHandler that does the same as for flood/gossipsub, except for these peers.

I would avoid doing this as it would complicate gossiping and potentially make the entire solution harder to evolve. Keeping protocols isolated has the advantage of allowing them to evolve at their own pace and keeps the entire thing more manageable. On top of that, pubsub as it is currently specified in libp2p, was never intended to support any sort of "light" node functionality and this might mean that any such extension will be hard to integrate.

So like mentioned by @kdeme I'm also of the opinion that this two functional parts should be kept independent from each other. Essentially there are two protocols to get/interact with messages:

- The pubsub/gossip that allows near real time message propagation
- An on demand RPC to retrieve historical messages

Light nodes only implement the on demand protocol and potentially only the client functionality, whereas full nodes support both gossiping and the on demand protocol (both the server and the client parts).

Reply:
> is something that pubsub already can do with the validators, furthermore, validators receive the entire message object, not just the data field, so it's possible to use additional fields such as from and sequence, to filter messages (might not be needed for waku).

But this would be a validation that checks a message before relaying to all peers, right? Not something selective for a set of peers?



> I would avoid doing this as it would complicate gossiping and potentially make the entire solution harder to evolve. Keeping protocols isolated has the advantage of allowing them to evolve at their own pace and keeps the entire thing more manageable.

I very much agree on that. Seem to make things much more complex, and will probably break on the next xSub. Plus, you're passing peer information from one domain (req/response protocol, which provides contentFilters) to another (PubSub), not sure if that is allowed, and if it is, doesn't really sound like the right way. I also don't see any benefits to the approach, the "full node" will still learn of all the topics (contentFilter) that the light node is interested in.

Reply:
Anyhow, it can also be that I misinterpreted that sentence and that `relaying` there means the relaying between the PubSub domain and the req/response. Either way, it means that the protocol is still incomplete.

Reply:
> But this would be a validation that checks a message before relaying to all peers, right? Not something selective for a set of peers?

Yes, this isn't selective it will run on all the peers that are participating in the topic/message.

Reply:
A lot of text :sweat_smile:. The specific section in the spec might need to be clarified to avoid misunderstandings. It is intended to be a separate and isolated protocol that light clients can implement. I believe @decanus is still looking into the specifics of how the messages will be passed on to the other node.

Reply:
I re-read the thread above after thinking about this some more, and just want to say thanks @dryajov and @kdeme for the insightful discussion.